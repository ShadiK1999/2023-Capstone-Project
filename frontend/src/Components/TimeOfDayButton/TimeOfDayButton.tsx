import { FC, useCallback } from 'react';
import { LabelBox, PlusMinusButton } from '../';

const MINUTES_INCREMENT = 15;

export type Time = {
  hour: number;
  minute: number;
};

type ComponentProps = {
  startTime: Time;
  endTime: Time;
  onUpdateStartTime: (time: Time) => void;
  onUpdateEndTime: (time: Time) => void;
};

const TimeOfDayButton: FC<ComponentProps> = ({
  startTime,
  endTime,
  onUpdateStartTime,
  onUpdateEndTime,
}) => {
  const handleUpdateStartTime = useCallback(
    (increment: number) => {
      let newStartMinute = startTime.minute + increment;
      let newStartHour = startTime.hour;

      if (newStartMinute > 59) {
        newStartHour = startTime.hour + 1;
        newStartMinute = 0;
      }

      if (newStartMinute < 0) {
        newStartHour = startTime.hour - 1;
        newStartMinute = 60 - MINUTES_INCREMENT;
      }

      if (
        newStartHour > endTime.hour ||
        (newStartHour === endTime.hour && newStartMinute > endTime.minute)
      ) {
        return;
      }

      if (newStartHour < 0) {
        return;
      }

      onUpdateStartTime({ hour: newStartHour, minute: newStartMinute });
    },
    [onUpdateStartTime, startTime, endTime],
  );

  const handleUpdateEndTime = useCallback(
    (increment: number) => {
      let newEndMinute = endTime.minute + increment;
      let newEndHour = endTime.hour;

      if (newEndMinute > 59) {
        newEndHour = endTime.hour + 1;
        newEndMinute = 0;
      }

      if (newEndMinute < 0) {
        newEndHour = endTime.hour - 1;
        newEndMinute = 60 - MINUTES_INCREMENT;
      }

      if (
        newEndHour < startTime.hour ||
        (newEndHour === startTime.hour && newEndMinute < startTime.minute)
      ) {
        return;
      }

      if (newEndHour > 23) {
        return;
      }

      onUpdateEndTime({ hour: newEndHour, minute: newEndMinute });
    },
    [onUpdateEndTime, startTime, endTime],
  );

  return (
    <>
      <LabelBox label="Open from">
        <PlusMinusButton
          leftButton={{
            onClick: () => handleUpdateStartTime(-MINUTES_INCREMENT),
          }}
          rightButton={{
            onClick: () => handleUpdateStartTime(MINUTES_INCREMENT),
          }}
        >
          {startTime.hour < 10 ? '0' + startTime.hour : startTime.hour}:
          {startTime.minute < 10 ? '0' + startTime.minute : startTime.minute}
        </PlusMinusButton>
      </LabelBox>
      <LabelBox label="Open until">
        <PlusMinusButton
          leftButton={{
            onClick: () => handleUpdateEndTime(-MINUTES_INCREMENT),
          }}
          rightButton={{
            onClick: () => handleUpdateEndTime(MINUTES_INCREMENT),
          }}
        >
          {endTime.hour < 10 ? '0' + endTime.hour : endTime.hour}:
          {endTime.minute < 10 ? '0' + endTime.minute : endTime.minute}
        </PlusMinusButton>
      </LabelBox>
    </>
  );
};

export { TimeOfDayButton };
