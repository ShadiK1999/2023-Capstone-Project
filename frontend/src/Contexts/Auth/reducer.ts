import { produce } from 'immer';
import { AuthState } from './index';

export type AuthAction =
  | {
      type: 'USER_LOGGED_IN';
      payload: string;
    }
  | {
      type: 'USER_LOGGED_OUT';
    };

export const authReducer = produce((draft: AuthState, action: AuthAction) => {
  switch (action.type) {
    case 'USER_LOGGED_IN':
      draft.jwt = action.payload;
      break;
    case 'USER_LOGGED_OUT':
      draft.jwt = null;
      break;
  }
});
