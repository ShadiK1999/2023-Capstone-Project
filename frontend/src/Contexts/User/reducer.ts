import { produce } from 'immer';
import { UserState } from '.';
import { TrustedNeighbour } from '../../Models/TrustedNeighbour';
import { TrustedPoint } from '../../Models/DeliveryPoint';

export type UserAction =
  | {
      type: 'FETCHED_USER';
      payload: TrustedNeighbour;
    }
  | {
      type: 'UPDATED_ADDRESS';
      payload: TrustedPoint;
    };

export const userReducer = produce((draft: UserState, action: UserAction) => {
  switch (action.type) {
    case 'FETCHED_USER':
      draft.user = action.payload;
      break;
    case 'UPDATED_ADDRESS':
      if (draft.user !== 'not-loaded') {
        draft.user.deliveryPoint = action.payload;
      }
      break;
  }
});
