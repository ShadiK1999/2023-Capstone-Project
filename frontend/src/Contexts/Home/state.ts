import { Dispatch, createContext } from 'react';
import { HomeAction } from './reducer';
import { DisplayPoint } from '../../Models/DeliveryPoint';

type IsochroneRings = {
  inner: google.maps.LatLngLiteral[];
  middle: google.maps.LatLngLiteral[];
  outer: google.maps.LatLngLiteral[];
};

type Map = {
  currentLocation: google.maps.LatLngLiteral;
  isochroneRings: IsochroneRings | null;
};

export type DayOfWeekIndex = 0 | 1 | 2 | 3 | 4 | 5 | 6;

export type Filter = {
  type: 'distance' | 'duration';
  mode: 'WALKING' | 'DRIVING' | 'CYCLING';
  value: number;
  dayOfWeek: DayOfWeekIndex;
  startHour: number;
  startMin: number;
  endHour: number;
  endMin: number;
};

export type HomeState = {
  map: Map;
  deliveryPoints: DisplayPoint[] | null;
  isLoading: boolean;
  filter: Filter;
};

export const initialState: HomeState = {
  map: {
    currentLocation: {
      lat: -37.8124,
      lng: 144.9623,
    },
    isochroneRings: null,
  },
  deliveryPoints: null,
  isLoading: false,
  filter: {
    type: 'duration',
    mode: 'WALKING',
    value: 5,
    dayOfWeek: 0,
    startHour: 9,
    startMin: 0,
    endHour: 17,
    endMin: 0,
  },
};

export const HomeContext = createContext<HomeState>(initialState);
export const HomeDispatchContext = createContext<Dispatch<HomeAction>>(() => void 0);
