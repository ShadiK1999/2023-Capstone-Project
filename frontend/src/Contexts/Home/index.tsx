import { FC, ReactNode, useReducer } from 'react';
import { HomeContext, HomeDispatchContext, initialState } from './state';
import { homeReducer } from './reducer';

type ComponentProps = {
  children: ReactNode;
};

const HomeProvider: FC<ComponentProps> = ({ children }) => {
  const [state, dispatch] = useReducer(homeReducer, initialState);

  return (
    <HomeContext.Provider value={state}>
      <HomeDispatchContext.Provider value={dispatch}>{children}</HomeDispatchContext.Provider>
    </HomeContext.Provider>
  );
};

export { HomeProvider };
