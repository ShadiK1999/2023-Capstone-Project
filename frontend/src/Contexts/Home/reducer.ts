import { HomeState, Filter } from './state';
import { IsochroneRingsOutput } from '../../Services/Isochrone';
import { produce } from 'immer';
import { DisplayPoint } from '../../Models/DeliveryPoint';

export type HomeAction =
  | {
      type: 'ISOCHRONE_READY';
      payload: IsochroneRingsOutput;
    }
  | {
      type: 'ADDRESS_CHANGED';
      payload: google.maps.LatLngLiteral;
    }
  | {
      type: 'FETCH_DELIVERY_POINTS_SUCCESS';
      payload: DisplayPoint[];
    }
  | {
      type: 'START_LOADING';
    }
  | {
      type: 'LOAD_SUCCESSFUL';
    }
  | {
      type: 'FILTER_INPUT_CHANGED';
      payload: Filter;
    };

export const homeReducer = produce((draft: HomeState, action: HomeAction) => {
  switch (action.type) {
    case 'ISOCHRONE_READY':
      draft.map.isochroneRings = action.payload;
      break;
    case 'ADDRESS_CHANGED':
      draft.map.currentLocation = action.payload;
      break;
    case 'FILTER_INPUT_CHANGED':
      draft.filter = action.payload;
      break;
    case 'FETCH_DELIVERY_POINTS_SUCCESS':
      draft.deliveryPoints = action.payload;
      break;
    case 'START_LOADING':
      draft.isLoading = true;
      break;
    case 'LOAD_SUCCESSFUL':
      draft.isLoading = false;
      break;
  }
});
