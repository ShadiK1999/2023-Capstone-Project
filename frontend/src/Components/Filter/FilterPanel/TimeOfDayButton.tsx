import { FC, useContext } from 'react';
import * as style from './TimeOfDayButton.styles';
import { HomeContext, HomeDispatchContext } from '../../../Contexts/Home/state';
import { LabelBox } from '../LabelBox';
import { PlusMinusButton } from '../PlusMinusButton';
import { produce } from 'immer';

const TimeOfDayButton: FC = () => {
  const dispatch = useContext(HomeDispatchContext);
  const { filter } = useContext(HomeContext);
  const hoursDisplay: string[] = [
    '12am',
    '1am',
    '2am',
    '3am',
    '4am',
    '5am',
    '6am',
    '7am',
    '8am',
    '9am',
    '10am',
    '11am',
    '12pm',
    '1pm',
    '2pm',
    '3pm',
    '4pm',
    '5pm',
    '6pm',
    '7pm',
    '8pm',
    '9pm',
    '10pm',
    '11pm',
  ];
  const MIN_HOUR = 0;
  const MAX_HOUR = 23;

  const handleUpdateOpenHours = (inputOpenTime: number) => {
    if (inputOpenTime >= filter.closeTime || inputOpenTime < MIN_HOUR) {
      return;
    }

    const updatedFilter = produce(filter, (draft: typeof filter) => {
      draft.openTime = inputOpenTime;
    });

    dispatch({
      type: 'FILTER_INPUT_CHANGED',
      payload: updatedFilter,
    });
  };

  const handleUpdateCloseHours = (inputCloseTime: number) => {
    if (inputCloseTime <= filter.openTime || inputCloseTime > MAX_HOUR) {
      return;
    }

    const updatedFilter = produce(filter, (draft: typeof filter) => {
      draft.closeTime = inputCloseTime;
    });

    dispatch({
      type: 'FILTER_INPUT_CHANGED',
      payload: updatedFilter,
    });
  };

  return (
    <div className={style.container}>
      <LabelBox label="Open from...">
        <PlusMinusButton
          leftButton={{
            onClick: () => handleUpdateOpenHours(filter.openTime - 1),
          }}
          rightButton={{
            onClick: () => handleUpdateOpenHours(filter.openTime + 1),
          }}
        >
          {hoursDisplay[filter.openTime]}
        </PlusMinusButton>
      </LabelBox>
      <LabelBox label="Until...">
        <PlusMinusButton
          leftButton={{
            onClick: () => handleUpdateCloseHours(filter.closeTime - 1),
          }}
          rightButton={{
            onClick: () => handleUpdateCloseHours(filter.closeTime + 1),
          }}
        >
          {hoursDisplay[filter.closeTime]}
        </PlusMinusButton>
      </LabelBox>
    </div>
  );
};

export { TimeOfDayButton };
