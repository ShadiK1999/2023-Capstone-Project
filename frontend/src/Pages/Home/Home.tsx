import { Map, InputBar, SideBar } from '../../Components';
import { HomeProvider } from '../../Contexts';

import * as styles from './Home.styles';

const Home = () => {
  return (
    <HomeProvider>
      <PageInternal />
    </HomeProvider>
  );
};

const PageInternal = () => {
  return (
    <div className={styles.Container}>
      <InputBar />
      <div></div>
      <Map />
      <SideBar />
    </div>
  );
};

export { Home };
