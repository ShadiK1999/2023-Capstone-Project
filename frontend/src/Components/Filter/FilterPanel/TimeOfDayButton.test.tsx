import { render, fireEvent } from '@testing-library/react';
import { ReactNode, useReducer } from 'react';
import { TimeOfDayButton } from './TimeOfDayButton';
import { homeReducer } from '../../../Contexts/Home/reducer';

import { describe, it, expect } from 'vitest';
import {
  HomeContext,
  Filter,
  HomeDispatchContext,
  HomeState,
  initialState,
} from '../../../Contexts/Home/state';

const mockedFilter: Filter = {
  type: 'duration',
  mode: 'WALKING',
  value: 15,
  daysOfWeek: ['SUN'],
  openTime: 8, // 8am
  closeTime: 10, // 10am
};

const mockedState: HomeState = {
  ...initialState,
  filter: mockedFilter,
};

type ComponentProps = {
  children: ReactNode;
};

function TestWrapper({ children }: ComponentProps) {
  const [state, dispatch] = useReducer(homeReducer, mockedState);

  return (
    <HomeContext.Provider value={state}>
      <HomeDispatchContext.Provider value={dispatch}>{children}</HomeDispatchContext.Provider>
    </HomeContext.Provider>
  );
}

describe('Matches DOM Snapshot', () => {
  it('Matches DOM Snapshot', () => {
    const tree = render(<TimeOfDayButton />);
    expect(tree).toMatchSnapshot();
  });

  it('open time cannot exceed close time', async () => {
    const { getAllByAltText, getAllByText } = render(
      <TestWrapper>
        <TimeOfDayButton />
      </TestWrapper>,
    );
    const plusButtons = getAllByAltText('PLUS');
    const plusOpeningHoursButton = plusButtons[0];

    const minusButtons = getAllByAltText('MINUS');
    const minusOpeningHoursButton = minusButtons[0];

    expect(getAllByText('8am')).toHaveLength(1);
    expect(getAllByText('10am')).toHaveLength(1);

    fireEvent.click(plusOpeningHoursButton);

    expect(getAllByText('9am')).toHaveLength(1);
    expect(getAllByText('10am')).toHaveLength(1);

    fireEvent.click(plusOpeningHoursButton);

    // Should remain at 9 am because 10am is the close time.
    expect(getAllByText('9am')).toHaveLength(1);
    expect(getAllByText('10am')).toHaveLength(1);

    fireEvent.click(minusOpeningHoursButton);

    expect(getAllByText('8am')).toHaveLength(1);
    expect(getAllByText('10am')).toHaveLength(1);
  });

  it('close time cannot exceed close time', async () => {
    const { getAllByAltText, getAllByText } = render(
      <TestWrapper>
        <TimeOfDayButton />
      </TestWrapper>,
    );
    const plusButtons = getAllByAltText('PLUS');
    const plusClosingHoursButton = plusButtons[1];

    const minusButtons = getAllByAltText('MINUS');
    const minusClosingHoursButton = minusButtons[1];

    expect(getAllByText('8am')).toHaveLength(1);
    expect(getAllByText('10am')).toHaveLength(1);

    fireEvent.click(minusClosingHoursButton);

    expect(getAllByText('8am')).toHaveLength(1);
    expect(getAllByText('9am')).toHaveLength(1);

    fireEvent.click(minusClosingHoursButton);

    // Should remain at 9am because 8am is the open time.
    expect(getAllByText('8am')).toHaveLength(1);
    expect(getAllByText('9am')).toHaveLength(1);

    fireEvent.click(plusClosingHoursButton);

    expect(getAllByText('8am')).toHaveLength(1);
    expect(getAllByText('10am')).toHaveLength(1);
  });
});
