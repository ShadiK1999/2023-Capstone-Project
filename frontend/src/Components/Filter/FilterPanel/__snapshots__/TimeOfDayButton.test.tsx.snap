// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Matches DOM Snapshot > Matches DOM Snapshot 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="f9vbxna"
      >
        <div
          class="flg5ft8"
        >
          <p>
            Open from...
          </p>
          <div
            class="fk0ntar"
          >
            <div
              class="f110hs72"
            >
              <div
                class="ffvwc24"
              >
                <img
                  alt="MINUS"
                  src="/src/Components/Filter/PlusMinusButton/Assets/Minus.svg"
                />
              </div>
              <div
                class="f1sz7xol"
              >
                12am
              </div>
              <div
                class="f1g8elpx"
              >
                <img
                  alt="PLUS"
                  src="/src/Components/Filter/PlusMinusButton/Assets/Plus.svg"
                />
              </div>
            </div>
          </div>
        </div>
        <div
          class="flg5ft8"
        >
          <p>
            Until...
          </p>
          <div
            class="fk0ntar"
          >
            <div
              class="f110hs72"
            >
              <div
                class="ffvwc24"
              >
                <img
                  alt="MINUS"
                  src="/src/Components/Filter/PlusMinusButton/Assets/Minus.svg"
                />
              </div>
              <div
                class="f1sz7xol"
              >
                11pm
              </div>
              <div
                class="f1g8elpx"
              >
                <img
                  alt="PLUS"
                  src="/src/Components/Filter/PlusMinusButton/Assets/Plus.svg"
                />
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="f9vbxna"
    >
      <div
        class="flg5ft8"
      >
        <p>
          Open from...
        </p>
        <div
          class="fk0ntar"
        >
          <div
            class="f110hs72"
          >
            <div
              class="ffvwc24"
            >
              <img
                alt="MINUS"
                src="/src/Components/Filter/PlusMinusButton/Assets/Minus.svg"
              />
            </div>
            <div
              class="f1sz7xol"
            >
              12am
            </div>
            <div
              class="f1g8elpx"
            >
              <img
                alt="PLUS"
                src="/src/Components/Filter/PlusMinusButton/Assets/Plus.svg"
              />
            </div>
          </div>
        </div>
      </div>
      <div
        class="flg5ft8"
      >
        <p>
          Until...
        </p>
        <div
          class="fk0ntar"
        >
          <div
            class="f110hs72"
          >
            <div
              class="ffvwc24"
            >
              <img
                alt="MINUS"
                src="/src/Components/Filter/PlusMinusButton/Assets/Minus.svg"
              />
            </div>
            <div
              class="f1sz7xol"
            >
              11pm
            </div>
            <div
              class="f1g8elpx"
            >
              <img
                alt="PLUS"
                src="/src/Components/Filter/PlusMinusButton/Assets/Plus.svg"
              />
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
